pulsar-broker/src/main/java/org/apache/pulsar/broker/intercept/BrokerInterceptor.java

```diff
@@ -58,6 +65,66 @@ default void beforeSendMessage(Subscription subscription,
                                    MessageMetadata msgMetadata) {
     }
 
+    /**
+     * Called by the broker when a new connection is created.
+     */
+    default void onConnectionCreated(ServerCnx cnx){
+    }
+
+    /**
+     * Called by the broker when a new connection is created.
+     */
+    default void producerCreated(ServerCnx cnx, Producer producer,
+                                 Map<String, String> metadata){
+    }
+
+    /**
+     * Intercept after a consumer is created.
+     *
+     * @param cnx client Connection
+     * @param consumer Consumer object
+     * @param metadata A map of metdata
+     */
+    default void consumerCreated(ServerCnx cnx,
+                                 Consumer consumer,
+                                 Map<String, String> metadata) {
+    }
+
+    /**
+     * Intercept after a message is produced.
+     *
+     * @param cnx client Connection
+     * @param producer Producer object
+     * @param publishContext Publish Context
+     */
+    default void messageProduced(ServerCnx cnx, Producer producer, long startTimeNs, long ledgerId,
```

Comment:
Could you please provide more context about why we should add `rateIn` as a param for this method?

Reply:
This is currently used in our logs to indicate the overall publish rate for the topic seen by the broker when this message was stored in the ledger. Used to troubleshoot latency issues and to support customer queries. Is there anything wrong in exposing this parameter?

Reply:
I see, I think it's better to get the rateIn from the producer?

Reply:
I agree. It was an oversight. I used rateIn as it was readily computed and available in MessagePublishContext::run. I will fix it now

Reply:
Fix done